app/lib/katello/repo_discovery.rb

```diff
@@ -42,31 +42,33 @@ def run(resume_point)
     private
 
     def docker_search
-      params = {
-        :accept => :json
-      }
-      params[:user] = @upstream_username unless @upstream_username.empty?
-      params[:password] = @upstream_password unless @upstream_password.empty?
-
-      RestClient.proxy = nil
-
+      proxy_uri = nil
       if @proxy && @proxy[:proxy_host]
         proxy_uri = URI(@proxy[:proxy_host])
         proxy_uri.user = @proxy[:proxy_user] if @proxy[:proxy_user]
         proxy_uri.password = @proxy[:proxy_password] if @proxy[:proxy_password]
         proxy_uri.port = @proxy[:proxy_port] if @proxy[:proxy_port]
-
-        RestClient.proxy = proxy_uri.to_s
       end
 
+      request_params = {
+        method: :get,
+        headers: { accept: :json },
+        url: "#{@uri}v1/search?q=#{@search}"
+      }
+
+      request_params[:headers][:user] = @upstream_username unless @upstream_username.empty?
+      request_params[:headers][:password] = @upstream_password unless @upstream_password.empty?
+      request_params[:proxy] = proxy_uri.to_s if proxy_uri
```

Comment:
when using a proxy password with special characters "@password#@"   i get an error on repo discovery:

Discovery failed. Error: bad password component

Its likely because rest client doesn't support this natively, but i think https://github.com/theforeman/foreman/commit/a196afb125a1af710b69bafd68d1803f7b630620 was added to help support this, we just have to escape it.  I believe that the upstream subscription stuff uses a helper method: 

https://github.com/Katello/katello/blob/master/app/lib/katello/resources/candlepin.rb#L99

 defined here:
https://github.com/Katello/katello/blob/master/app/lib/katello/util/http_proxy.rb#L4

Reply:
i only checked this because literally everytime we add proxy support to something, this break ;)

Reply:
also this happens on master too, but probably good to fix here